### PR TITLE
[SavedPlans] Check if the User has already saved the needed Plan, thus do Not activate Save button

### PR DESCRIPTION
**Pull request by sananand007**
## Description
Currently if we have saved Plans for the user Profile, we do not auto detect them and act on those. 
With this change
- we will check for saved plans for the User immediately when user logs in and does a search 
- for most cases the user will be generating original or unique plans based on date and other search criteria
- if for the particular search the plan is not unique and is already been saved, the `save` button will be inactive, stating that the plan is already saved 

## Solution
* Describe your code changes in detail for reviewers as well as documentation purposes. 
* For bug fixes, explain the solutions and how they fix the issues.
* For new features, describe the high-level ideas.

## Testing
- [ ] Integration testing on Heroku staging
- [ ] Added new unit tests

## Checks
- [ ] Have you removed commented code?
- [ ] Have you used gofmt to format your code?

### CODE REVIEW

File: planner/planner.go [hunk 1]
```diff
@@ -572,6 +573,21 @@ func (p *MyPlanner) getPlanningApi(ctx *gin.Context) {
 		return
 	}
 
+	if userView.IsSavedPlans {
```
Comment (timwangmusic) [inline planner/planner.go:576]: Why we need to check this Boolean?

File: planner/planner.go [hunk 2]
```diff
@@ -572,6 +573,21 @@ func (p *MyPlanner) getPlanningApi(ctx *gin.Context) {
...
+		var s1 []string = make([]string, 0)
+		var numOfPlanResultsAvail int = min(numResultsInt, len(planningResp.TravelPlans))
```
Comment (timwangmusic) [inline planner/planner.go:578]: Use the constant instead of a string literal

File: planner/planner.go [hunk 3]
```diff
@@ -572,6 +573,21 @@ func (p *MyPlanner) getPlanningApi(ctx *gin.Context) {
...
+		if err != nil {
+			logger.Debugf("Cannot find plan with key %s: %v", savedUserOriginalPlans, err)
+		}
+		for PlanIndex := 0; PlanIndex < numOfPlanResultsAvail; PlanIndex++ {
+			var SearchFoundNewPlanId = strings.Join([]string{"travel_plan", planningResp.TravelPlans[PlanIndex].ID}, ":")
+			if isPlanIDPresentInSavedList(SearchFoundNewPlanId, s1) {
+				planningResp.TravelPlans[PlanIndex].Saved = true
+			}
+		}
+	}
```
Comment (timwangmusic) [inline planner/planner.go:590]: Try out these in a separate function

File: iowrappers/users.go [hunk 4]
```diff
@@ -327,6 +329,21 @@ func (r *RedisClient) SaveUserPlan(context context.Context, userView user.View,
 	}
 
 	redisKey := strings.Join([]string{UserSavedTravelPlanPrefix, "user", userView.ID, "plan", planView.ID}, ":")
+	Logger.Debugf("-> Saving the Plan %s for user %s here", planView.ID, userView.ID)
+
+	SavedPlanState := strings.Join([]string{UserSavedPlanState, "user", userView.ID}, ":")
```
Comment (timwangmusic) [inline iowrappers/users.go:334]: I don't think we need this.

**Review by timwangmusic — COMMENTED** [review]
Good work! Left some comments.

File: iowrappers/redis_client.go [hunk 5]
```diff
@@ -701,3 +701,12 @@ func (r *RedisClient) FetchSingleRecord(context context.Context, redisKey string
 	}
 	return nil
 }
+
+func (r *RedisClient) FetchSingleRecordTypeSet(context context.Context, redisKey string, response interface{}) ([]string, error) {
```
[the remark this answers is not in the document]
Comment (sananand007) [reply iowrappers/redis_client.go:705]: yes I can use that within `redis_client.go` instead. But if I need to access `redisClient` api's within `planner` I would have to do `"github.com/redis/go-redis/v9"` , that might be too much ? @timwangmusic ?

[the remark this answers is not in the document]
Comment (timwangmusic) [reply iowrappers/redis_client.go:705]: We are already using go-Redis.get the client use client.Get method

File: planner/planner.go (hunk 1 above)
Comment (sananand007) [reply planner/planner.go]: dont need this anymore.

File: planner/planner.go [hunk 6]
```diff
@@ -476,6 +478,24 @@ func (p *MyPlanner) getOptimalPlan(ctx *gin.Context) {
 	}
 }
 
+func (p *MyPlanner) UpdatePlanSavedStateForUser(ctx *gin.Context, numResultsInt int, resp PlanningResponse, uv user.View, lgr *zap.SugaredLogger) {
```
Comment (timwangmusic) [inline planner/planner.go:481]: use `logger := iowrappers.Logger` instead of passing it into the function.

File: planner/planner.go [hunk 7]
```diff
@@ -476,6 +478,24 @@ func (p *MyPlanner) getOptimalPlan(ctx *gin.Context) {
...
+	var s1 = make([]string, 0)
+	var err error
+	var numOfPlanResultsAvail int = min(numResultsInt, len(resp.TravelPlans))
+	var savedUserOriginalPlans = strings.Join([]string{"user_saved_travel_plans", "user", uv.ID, "plans"}, ":")
+	s1, err = p.RedisClient.FetchSingleRecordTypeSet(ctx, savedUserOriginalPlans, &s1)
+	if err != nil {
```
Comment (timwangmusic) [inline planner/planner.go:487]: return `err` to caller.

File: planner/planner.go [hunk 8]
```diff
@@ -476,6 +478,24 @@ func (p *MyPlanner) getOptimalPlan(ctx *gin.Context) {
 	}
 }
 
+func (p *MyPlanner) UpdatePlanSavedStateForUser(ctx *gin.Context, numResultsInt int, resp PlanningResponse, uv user.View, lgr *zap.SugaredLogger) {
+	var s1 = make([]string, 0)
+	var err error
+	var numOfPlanResultsAvail int = min(numResultsInt, len(resp.TravelPlans))
+	var savedUserOriginalPlans = strings.Join([]string{"user_saved_travel_plans", "user", uv.ID, "plans"}, ":")
+	s1, err = p.RedisClient.FetchSingleRecordTypeSet(ctx, savedUserOriginalPlans, &s1)
```
Comment (timwangmusic) [inline planner/planner.go:486]: Style nit: `s1, err :=` so you don't need to declare them separately.

File: planner/planner.go [hunk 9]
```diff
@@ -476,6 +478,24 @@ func (p *MyPlanner) getOptimalPlan(ctx *gin.Context) {
 	}
 }
 
+func (p *MyPlanner) UpdatePlanSavedStateForUser(ctx *gin.Context, numResultsInt int, resp PlanningResponse, uv user.View, lgr *zap.SugaredLogger) {
+	var s1 = make([]string, 0)
+	var err error
+	var numOfPlanResultsAvail int = min(numResultsInt, len(resp.TravelPlans))
+	var savedUserOriginalPlans = strings.Join([]string{"user_saved_travel_plans", "user", uv.ID, "plans"}, ":")
```
Comment (timwangmusic) [inline planner/planner.go:485]: There is a constant in `iowrappers` for `user_saved_travel_plans`

File: planner/planner.go [hunk 10]
```diff
@@ -1044,3 +1065,12 @@ func getPlaceIcon(placeTypes []POI.PlaceCategory, pIdx int) string {
 	}
 	return string(placeTypeToIcon[placeTypes[pIdx]])
 }
+
+func isPlanIDPresentInSavedList(planID string, savedPlaces []string) bool {
```
Comment (timwangmusic) [inline planner/planner.go:1069]: Are the parameter names correct? Should the `savedPlaces` be `userSavedPlans`?

File: planner/users.go [hunk 11]
```diff
@@ -167,7 +167,6 @@ func (p *MyPlanner) UserAuthentication(ctx *gin.Context, minimumUserLevel user.L
 	if findUserErr != nil {
 		return userView, findUserErr
 	}
-
```
Comment (timwangmusic) [inline planner/users.go:170]: Please do not remove irrelevant lines.

File: iowrappers/users.go [hunk 12]
```diff
@@ -327,6 +327,7 @@ func (r *RedisClient) SaveUserPlan(context context.Context, userView user.View,
 	}
 
 	redisKey := strings.Join([]string{UserSavedTravelPlanPrefix, "user", userView.ID, "plan", planView.ID}, ":")
+
```
Comment (timwangmusic) [inline iowrappers/users.go:330]: remove this change please.

File: planner/planner.go [hunk 13]
```diff
@@ -476,6 +477,25 @@ func (p *MyPlanner) getOptimalPlan(ctx *gin.Context) {
...
 }
 
+func (p *MyPlanner) UpdatePlanSavedStateForUser(ctx *gin.Context, numResultsInt int, resp PlanningResponse, uv user.View) error {
+	var err error
+	logger := iowrappers.Logger
+	var numOfPlanResultsAvail int = min(numResultsInt, len(resp.TravelPlans))
+	var savedUserOriginalPlans = strings.Join([]string{iowrappers.UserSavedTravelPlansPrefix, "user", uv.ID, "plans"}, ":")
+	s1, err := p.RedisClient.FetchSingleRecordTypeSet(ctx, savedUserOriginalPlans)
+	if err != nil {
+		logger.Debugf("Cannot find plan with key %s: %v", savedUserOriginalPlans, err)
```
Comment (timwangmusic) [inline planner/planner.go:487]: we do not need to log here if the error is handled outside the function.

File: planner/planner.go [hunk 14]
```diff
@@ -476,6 +477,25 @@ func (p *MyPlanner) getOptimalPlan(ctx *gin.Context) {
...
+		return err
+	}
+	for PlanIndex := 0; PlanIndex < numOfPlanResultsAvail; PlanIndex++ {
+		var SearchFoundNewPlanId = strings.Join([]string{"travel_plan", resp.TravelPlans[PlanIndex].ID}, ":")
```
Comment (timwangmusic) [inline planner/planner.go:491]: This variable name is confusing, just use something like `targetPlanId`.

File: planner/planner.go [hunk 15]
```diff
@@ -476,6 +477,25 @@ func (p *MyPlanner) getOptimalPlan(ctx *gin.Context) {
 	}
 }
 
+func (p *MyPlanner) UpdatePlanSavedStateForUser(ctx *gin.Context, numResultsInt int, resp PlanningResponse, uv user.View) error {
+	var err error
+	logger := iowrappers.Logger
+	var numOfPlanResultsAvail int = min(numResultsInt, len(resp.TravelPlans))
+	var savedUserOriginalPlans = strings.Join([]string{iowrappers.UserSavedTravelPlansPrefix, "user", uv.ID, "plans"}, ":")
+	s1, err := p.RedisClient.FetchSingleRecordTypeSet(ctx, savedUserOriginalPlans)
```
Comment (timwangmusic) [inline planner/planner.go:485]: rename `s1` to `savedPlanIds`.

**Review by timwangmusic — APPROVED** [review]
I have some minor comments, but generally good look :). Please deploy to testing.

**Pull request merged**